### ISSUE TITLE
Thumbor presets - what's the right way to implement them?

### ISSUE THREAD

**Issue by alexmorozov** [thumbor/thumbor#662]
Hi guys, since I haven't found a mailing list or something, I'll write it here, hope it's OK.

So what I want is to have some predefined "presets", like, you know "crop a pic 200x200 smart, then overlay a watermark, round the corners, bla-bla-bla", and I want to be able to apply that presets by the name without having to specify all that filters and stuff in URL. Actually I've already put together some [draft code](https://github.com/alexmorozov/thumbor-presets), and I'd love to hear your opinion on how to do a redirect from my "preset" url to the actual imaging handler. For now I just make another HTTP request and throw it in a router, but it seems a bit harsh and redundant to me. On the other side, this guarantees (to some extent) a compatibility even if the underlying thumbor code changes.

So what's your opinion, is there another, more clean way to implement such kind of functionality?

And yes, thanks for the amazing software, you rock.

**Comment by guilhermef** [comment]
@alexmorozov Don't you think it's better to have these presets on libthumbor, and then use the https://github.com/thumbor-community/shortener ?

**Comment by alexmorozov** [comment]
@guilhermef , thanks a lot for a pointer, didn't know about that module. Seems very interesting, but it solves a slightly different task and uses an additional storage, which doesn't sound that cool to me. 

Anyway, I've already had a sneak peek at the `shortener` source code, and the idea of inheriting a handler from the `ImagingHandler` is just brilliant. Will give it a try. By the way, does making an HTTP request inside a Tornado handler have any consequences and is it generally a bad thing?

**Comment by masom** [comment]
@alexmorozov Something we did for PopKey is to overload the url handling with custom code that does unwrap into a thumbor-compatible kwargs.

We had to load a different tornado app than what thumbor provides by default ( the `-a` option let's you do that). Take a look at https://github.com/thumbor-community/core on how this can be done.

Example:
`https://m.popkey.co/2a2e8e/46mJ6_s-200x150.gif`

The url sigining is shorter than the standard Thumbor sha-base64 and the size marking is part of the file name. There a lot of other magic going on to serve the image.

**Comment by masom** [comment]
Specifically: https://github.com/thumbor-community/core/blob/master/tc_core/app.py loads more handlers than the default thumbor app.

**Comment by alexmorozov** [comment]
@masom Yeah, thanks a ton, I've already figured out how to specify a different app with a custom handler. Actually I have some working code (see the first post), but the way it redirects a request to the stock `ImageHandler` seems a bit stupid to me. That reminds me, do you store your signature-url mappings in a intermediate storage, or you have a direct idempotent (I'm pretty sure I'm misusing the word, but hope you get the point) mapping?

**Comment by masom** [comment]
Specifics to popkey:

- We re-implemented the url parsing scheme to support something different: https://github.com/thumbor/thumbor/blob/master/thumbor/url.py#L14 and we created a new class overloading the ImageHandler to first check the different possible urls. Basically having two lines for: https://github.com/thumbor/thumbor/blob/master/thumbor/app.py#L50

You basically want a handler that implements
https://github.com/thumbor/thumbor/blob/master/thumbor/handlers/imaging.py#L86 and unpack the non-thumbor url into kwargs thumbor will understand.

Within the url parsing code we also implemented an SQLAlchemy connector that loads the image slug and finds the original path in a read-only database.
